drivers/nvme-queue.cc

```diff
@@ -0,0 +1,464 @@
+{
+    _sq_addr[_sq_tail] = *cmd;
+    advance_sq_tail();
+    mmio_setl(_sq_doorbell,_sq_tail);
+    return _sq_tail;
+}
+
+void nvme_queue_pair::wait_for_completion_queue_entries()
+{
+    sched::thread::wait_until([this] {
+        bool have_elements = this->completion_queue_not_empty();
+        if (!have_elements) {
+            this->enable_interrupts();
+            //check if we got a new cqe between completion_queue_not_empty()
+            //and enable_interrupts()
+            have_elements = this->completion_queue_not_empty();
+            if (have_elements) {
+                this->disable_interrupts();
+            }
+        }
+
+        trace_nvme_wait_for_completion_queue_entries(_driverid,_id,have_elements);
+        return have_elements;
+    });
+}
+
+int nvme_queue_pair::map_prps(u16 cid, void* data, u64 datasize, u64* prp1, u64* prp2)
+{
+    u64 addr = (u64) data;
+    *prp1 = addr;
+    *prp2 = 0;
+    int numpages = 0;
+    u64 offset = addr - ( (addr >> NVME_PAGESHIFT) << NVME_PAGESHIFT );
+    if(offset) numpages = 1;
+
+    numpages += ( datasize - offset + NVME_PAGESIZE - 1) >> NVME_PAGESHIFT;
+
+    if (numpages == 2) {
+        *prp2 = ((addr >> NVME_PAGESHIFT) +1 ) << NVME_PAGESHIFT;
+    } else if (numpages > 2) {
+        assert(numpages / 512 == 0);
+        u64* prplist = (u64*) memory::alloc_phys_contiguous_aligned(numpages * 8, 4096);
+        assert(prplist != nullptr);
+        *prp2 = mmu::virt_to_phys(prplist);
+        _prplists_in_use.at(cid / _qsize)[cid % _qsize] = prplist;
+        
+        addr = ((addr >> NVME_PAGESHIFT) +1 ) << NVME_PAGESHIFT;
+        prplist[0] = addr;
+
+        for (int i = 1; i < numpages - 1; i++) {
+            addr += NVME_PAGESIZE;
+            prplist[i] = addr;
+        }
+    }
+    return 0;
+}
+
+std::unique_ptr<nvme_cq_entry_t> nvme_queue_pair::get_completion_queue_entry()
+{
+    if(!completion_queue_not_empty()) {
+        return nullptr;
+    }
+
+    auto* tcqe = new nvme_cq_entry_t; 
+    *tcqe = _cq_addr[_cq_head];
+    std::unique_ptr<nvme_cq_entry_t> cqe(tcqe);
+    assert(cqe->p == _cq_phase_tag);
+
+    if(++_cq_head == _qsize) {
+        _cq_head -= _qsize;
+        _cq_phase_tag = !_cq_phase_tag;
+    }
+    return cqe; 
+}
+
+
+bool nvme_queue_pair::completion_queue_not_empty() const
+{
+    bool a = reinterpret_cast<volatile nvme_cq_entry_t*>(&_cq_addr[_cq_head])->p == _cq_phase_tag;
+    trace_nvme_completion_queue_not_empty(_driverid,_id,a);
+    return a;//_cq_addr[_cq_head].p == _cq_phase_tag;
+}
+
+void nvme_queue_pair::enable_interrupts()
+{
+    _dev->msix_unmask_entry(_id);
+    trace_nvme_enable_interrupts(_driverid,_id);
+}
+
+void nvme_queue_pair::disable_interrupts()
+{
+    _dev->msix_mask_entry(_id);
+    trace_nvme_disable_interrupts(_driverid,_id);
+}
+
+//only use with interrupts disabled
+std::unique_ptr<nvme_cq_entry_t> nvme_queue_pair::check_for_completion(u16 cid)
```

Comment:
The function isn't used currently. The function polls for cqe instead of relying on interrupts. I mostly used this while checking if the interrupts work correctly. 
I left it in since it might be useful to have an alternative to msix interrupts in case something goes wrong